src/lib/unlang/module.c

```diff
@@ -905,6 +986,44 @@ static unlang_action_t unlang_module(rlm_rcode_t *p_result, request_t *request,
 		goto done;
 	}
 
+	if (mc->method_env) {
+		/*
+		 *	We have a method env and this is the first call.
+		 *	Allocate the data structure.
+		 */
+		if (!state->last_expanded) {
```

Comment:
Could you do an enum/switch here just to make it really explicit what's going on, `MOD_ENV_EXP_INIT`, `MOD_ENV_EXP_DONE`, `MOD_CALL`...  Or you can split them into different functions.  I think that's what @alandekok ended up doing for the init code.